File: _legacy/api_gateway/routers/projects.py
"""
Project management endpoints for Agent-OS compatibility.
"""

from fastapi import APIRouter, HTTPException, Query
from typing import Optional, List, Dict, Any
from datetime import datetime, timedelta
import logging

from services.convex_client import get_convex_client, get_mock_convex_client
import os

logger = logging.getLogger(__name__)

router = APIRouter()


def transform_project_data(
    convex_project: Dict[str, Any],
    client_name: Optional[str] = None
) -> Dict[str, Any]:
    """
    Transform Convex project data to Agent-OS format.

    Args:
        convex_project: Project data from Convex
        client_name: Client name for the project

    Returns:
        Transformed project data for Agent-OS
    """
    # Calculate progress based on milestones or estimate
    progress = calculate_progress(convex_project.get("milestones", []))

    # Transform team members
    team = convex_project.get("teamMembers", [])
    if isinstance(team[0] if team else None, str):
        # If team members are just names/IDs, keep as is
        team_list = team
    else:
        # If team members are objects, extract names
        team_list = [member.get("name", member) for member in team]

    return {
        "id": convex_project.get("_id"),
        "name": convex_project.get("name"),
        "clientId": convex_project.get("clientId"),
        "clientName": client_name or "Unknown Client",
        "description": convex_project.get("description"),
        "status": convex_project.get("status"),
        "priority": convex_project.get("priority"),
        "type": convex_project.get("type"),
        "progress": progress,
        "dueDate": datetime.fromtimestamp(
            convex_project.get("dueDate", 0) / 1000
        ).isoformat() if convex_project.get("dueDate") else None,
        "startDate": datetime.fromtimestamp(
            convex_project.get("startDate", 0) / 1000
        ).isoformat() if convex_project.get("startDate") else None,
        "completedDate": datetime.fromtimestamp(
            convex_project.get("completedDate", 0) / 1000
        ).isoformat() if convex_project.get("completedDate") else None,
        "team": team_list,
        "budget": convex_project.get("budget", 0),
        "spent": convex_project.get("actualCost", 0),
        "milestones": convex_project.get("milestones", [])
    }


def calculate_progress(milestones: List[Dict[str, Any]]) -> int:
    """
    Calculate project progress based on milestones.

    Args:
        milestones: List of project milestones

    Returns:
        Progress percentage (0-100)
    """
    if not milestones:
        return 0

    completed = sum(1 for m in milestones if m.get("status") == "completed")
    total = len(milestones)

    return int((completed / total) * 100)


@router.get("")
async def get_projects(
    status: Optional[str] = Query(None, description="Filter by status"),
    clientId: Optional[str] = Query(None, description="Filter by client ID"),
    priority: Optional[str] = Query(None, description="Filter by priority"),
    type: Optional[str] = Query(None, description="Filter by project type"),
    sort: Optional[str] = Query("dueDate", description="Sort field"),
    order: Optional[str] = Query("asc", description="Sort order (asc/desc)"),
    page: int = Query(1, ge=1, description="Page number"),
    limit: int = Query(10, ge=1, le=100, description="Items per page")
) -> Dict[str, Any]:
    """
    Get all projects with optional filtering.

    Returns:
        List of projects in Agent-OS format
    """
    try:
        # Use mock client in development if Convex not configured
        if os.getenv("CONVEX_URL"):
            client = get_convex_client()
        else:
            client = get_mock_convex_client()

        # Build query arguments
        query_args = {}
        if status:
            query_args["status"] = status
        if clientId:
            query_args["clientId"] = clientId
        if priority:
            query_args["priority"] = priority
        if type:
            query_args["type"] = type

        # Query Convex
        projects = await client.query("projects:list", query_args)

        # Get client names for all projects
        client_ids = list(set(p.get("clientId") for p in projects if p.get("clientId")))
        clients = {}
        for cid in client_ids:
            client_data = await client.query("clients:get", {"id": cid})
            if client_data:
                clients[cid] = client_data.get("companyName", "Unknown")

        # Transform to Agent-OS format
        transformed_projects = []
        for convex_project in projects:
            client_name = clients.get(convex_project.get("clientId"), "Unknown Client")
            transformed_projects.append(
                transform_project_data(convex_project, client_name)
            )

        # Sort projects
        if sort:
            reverse = order == "desc"
            if sort == "dueDate":
                transformed_projects.sort(
                    key=lambda x: x.get("dueDate", ""),
                    reverse=reverse
                )
            elif sort == "priority":
                priority_order = {"urgent": 0, "high": 1, "medium": 2, "low": 3}
                transformed_projects.sort(
                    key=lambda x: priority_order.get(x.get("priority", "low"), 3),
                    reverse=reverse
                )
            elif sort == "progress":
                transformed_projects.sort(
                    key=lambda x: x.get("progress", 0),
                    reverse=reverse
                )

        # Apply pagination
        start_idx = (page - 1) * limit
        end_idx = start_idx + limit
        paginated_projects = transformed_projects[start_idx:end_idx]

        return {
            "success": True,
            "data": paginated_projects,
            "total": len(transformed_projects),
            "page": page,
            "pages": (len(transformed_projects) + limit - 1) // limit
        }

    except Exception as e:
        logger.error(f"Error fetching projects: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{project_id}")
async def get_project(project_id: str) -> Dict[str, Any]:
    """
    Get a single project by ID.

    Args:
        project_id: Project identifier

    Returns:
        Detailed project information
    """
    try:
        # Use mock client in development if Convex not configured
        if os.getenv("CONVEX_URL"):
            client = get_convex_client()
        else:
            client = get_mock_convex_client()

        # Query Convex for specific project
        convex_project = await client.query("projects:get", {"id": project_id})

        if not convex_project:
            raise HTTPException(status_code=404, detail="Project not found")

        # Get client information
        client_data = await client.query("clients:get", {"id": convex_project.get("clientId")})
        client_name = client_data.get("companyName") if client_data else "Unknown Client"

        # Get deliverables for this project
        deliverables = await client.query("deliverables:list", {"projectId": project_id})

        # Transform to detailed Agent-OS format
        transformed = transform_project_data(convex_project, client_name)

        # Add team details with roles
        if convex_project.get("teamMembers"):
            transformed["team"] = [
                {
                    "id": f"user_{idx}",
                    "name": member,
                    "role": "Developer"  # In production, fetch actual roles
                }
                for idx, member in enumerate(convex_project.get("teamMembers", []))
            ]

        # Add deliverables
        transformed["deliverables"] = [
            {
                "id": d.get("_id"),
                "name": d.get("name"),
                "status": d.get("status")
            }
            for d in deliverables
        ]

        # Calculate time tracked (mock calculation)
        if convex_project.get("startDate"):
            days_active = (
                datetime.now() - datetime.fromtimestamp(convex_project["startDate"] / 1000)
            ).days
            transformed["timeTracked"] = days_active * 6.5  # Mock hours calculation

        return {
            "success": True,
            "data": transformed
        }

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error fetching project {project_id}: {e}")
        raise HTTPException(status_code=500, detail=str(e))

File: _legacy/convex/projects.ts
import { v } from "convex/values";
import { mutation, query } from "./_generated/server";
import { Doc, Id } from "./_generated/dataModel";

// Create a new project
export const createProject = mutation({
  args: {
    clientId: v.id("clients"),
    name: v.string(),
    description: v.optional(v.string()),
    type: v.string(),
    priority: v.optional(v.string()),
    dueDate: v.optional(v.number()),
    budget: v.optional(v.number()),
    teamMembers: v.optional(v.array(v.string())),
  },
  handler: async (ctx, args) => {
    const now = Date.now();

    const projectId = await ctx.db.insert("projects", {
      ...args,
      status: "planning",
      priority: args.priority || "medium",
      startDate: now,
      createdAt: now,
      updatedAt: now,
    });

    // Get client info for notification
    const client = await ctx.db.get(args.clientId);

    // Create activity log
    await ctx.db.insert("activityLogs", {
      action: "created",
      entityType: "project",
      entityId: projectId,
      entityName: args.name,
      metadata: {
        clientName: client?.companyName,
      },
      timestamp: now,
    });

    // Notify team members
    if (args.teamMembers && args.teamMembers.length > 0) {
      for (const userId of args.teamMembers) {
        await ctx.db.insert("notifications", {
          userId,
          type: "info",
          category: "project",
          title: "New Project Assignment",
          message: `You've been assigned to project: ${args.name}`,
          isRead: false,
          actionUrl: `/dashboard/projects/${projectId}`,
          actionLabel: "View Project",
          entityType: "project",
          entityId: projectId,
          createdAt: now,
        });
      }
    }

    return projectId;
  },
});

// Get all projects with optional filtering
export const getProjects = query({
  args: {
    clientId: v.optional(v.id("clients")),
    status: v.optional(v.string()),
    priority: v.optional(v.string()),
    type: v.optional(v.string()),
    limit: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    let query = ctx.db.query("projects");

    if (args.clientId) {
      query = query.withIndex("by_client", (q) => q.eq("clientId", args.clientId));
    } else if (args.status) {
      query = query.withIndex("by_status", (q) => q.eq("status", args.status));
    } else if (args.priority) {
      query = query.withIndex("by_priority", (q) => q.eq("priority", args.priority));
    } else if (args.type) {
      query = query.withIndex("by_type", (q) => q.eq("type", args.type));
    }

    const projects = await query
      .order("desc")
      .take(args.limit || 50);

    // Enrich with client names and deliverable counts
    const enrichedProjects = await Promise.all(
      projects.map(async (project) => {
        const client = await ctx.db.get(project.clientId);
        const deliverables = await ctx.db
          .query("deliverables")
          .withIndex("by_project", (q) => q.eq("projectId", project._id))
          .collect();

        const completedDeliverables = deliverables.filter(d => d.status === "delivered").length;
        const progress = deliverables.length > 0
          ? Math.round((completedDeliverables / deliverables.length) * 100)
          : 0;

        return {
          ...project,
          clientName: client?.companyName || "Unknown",
          totalDeliverables: deliverables.length,
          completedDeliverables,
          progress,
          isOverdue: project.dueDate && project.dueDate < Date.now() && project.status !== "completed",
        };
      })
    );

    return enrichedProjects;
  },
});

// Get a single project by ID
export const getProject = query({
  args: { id: v.id("projects") },
  handler: async (ctx, args) => {
    const project = await ctx.db.get(args.id);
    if (!project) return null;

    // Get related data
    const client = await ctx.db.get(project.clientId);
    const deliverables = await ctx.db
      .query("deliverables")
      .withIndex("by_project", (q) => q.eq("projectId", args.id))
      .collect();

    const messages = await ctx.db
      .query("messages")
      .withIndex("by_project", (q) => q.eq("projectId", args.id))
      .order("desc")
      .take(10);

    const documents = await ctx.db
      .query("documents")
      .withIndex("by_project", (q) => q.eq("projectId", args.id))
      .collect();

    return {
      ...project,
      client,
      deliverables,
      recentMessages: messages,
      documents,
    };
  },
});

// Update project status
export const updateProjectStatus = mutation({
  args: {
    id: v.id("projects"),
    status: v.string(),
  },
  handler: async (ctx, args) => {
    const project = await ctx.db.get(args.id);
    if (!project) throw new Error("Project not found");

    const now = Date.now();
    const oldStatus = project.status;

    const updates: any = {
      status: args.status,
      updatedAt: now,
    };

    // If completing the project, set completion date
    if (args.status === "completed" && !project.completedDate) {
      updates.completedDate = now;
    }

    await ctx.db.patch(args.id, updates);

    // Log the status change
    await ctx.db.insert("activityLogs", {
      action: "updated",
      entityType: "project",
      entityId: args.id,
      entityName: project.name,
      changes: {
        status: { from: oldStatus, to: args.status },
      },
      timestamp: now,
    });

    // Notify client if project is completed
    if (args.status === "completed") {
      const client = await ctx.db.get(project.clientId);
      if (client) {
        await ctx.db.insert("notifications", {
          userId: client.primaryContact.email, // TODO: Get actual user ID
          type: "success",
          category: "project",
          title: "Project Completed",
          message: `Your project "${project.name}" has been completed!`,
          isRead: false,
          actionUrl: `/portal/projects/${args.id}`,
          actionLabel: "View Project",
          entityType: "project",
          entityId: args.id,
          createdAt: now,
        });
      }
    }

    return { success: true };
  },
});

// Update project details
export const updateProject = mutation({
  args: {
    id: v.id("projects"),
    name: v.optional(v.string()),
    description: v.optional(v.string()),
    priority: v.optional(v.string()),
    dueDate: v.optional(v.number()),
    budget: v.optional(v.number()),
    actualCost: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    const { id, ...updates } = args;
    const now = Date.now();

    await ctx.db.patch(id, {
      ...updates,
      updatedAt: now,
    });

    return { success: true };
  },
});

// Add milestone to project
export const addMilestone = mutation({
  args: {
    projectId: v.id("projects"),
    name: v.string(),
    dueDate: v.number(),
  },
  handler: async (ctx, args) => {
    const project = await ctx.db.get(args.projectId);
    if (!project) throw new Error("Project not found");

    const milestones = project.milestones || [];
    milestones.push({
      name: args.name,
      dueDate: args.dueDate,
      status: "pending",
    });

    await ctx.db.patch(args.projectId, {
      milestones,
      updatedAt: Date.now(),
    });

    return { success: true };
  },
});

// Update milestone status
export const updateMilestone = mutation({
  args: {
    projectId: v.id("projects"),
    milestoneName: v.string(),
    status: v.string(),
  },
  handler: async (ctx, args) => {
    const project = await ctx.db.get(args.projectId);
    if (!project || !project.milestones) throw new Error("Project or milestone not found");

    const now = Date.now();
    const milestones = project.milestones.map(m => {
      if (m.name === args.milestoneName) {
        return {
          ...m,
          status: args.status,
          completedDate: args.status === "completed" ? now : m.completedDate,
        };
      }
      return m;
    });

    await ctx.db.patch(args.projectId, {
      milestones,
      updatedAt: now,
    });

    return { success: true };
  },
});

// Get project statistics
export const getProjectStats = query({
  handler: async (ctx) => {
    const projects = await ctx.db.query("projects").collect();

    const stats = {
      total: projects.length,
      byStatus: {} as Record<string, number>,
      byType: {} as Record<string, number>,
      byPriority: {} as Record<string, number>,
      active: 0,
      overdue: 0,
      completedOnTime: 0,
      averageCompletionTime: 0,
      totalBudget: 0,
      totalActualCost: 0,
    };

    let totalCompletionTime = 0;
    let completedCount = 0;

    for (const project of projects) {
      // Count by status
      stats.byStatus[project.status] = (stats.byStatus[project.status] || 0) + 1;

      // Count by type
      stats.byType[project.type] = (stats.byType[project.type] || 0) + 1;

      // Count by priority
      stats.byPriority[project.priority] = (stats.byPriority[project.priority] || 0) + 1;

      // Count active
      if (project.status === "in_progress") {
        stats.active++;
      }

      // Check overdue
      if (project.dueDate && project.dueDate < Date.now() && project.status !== "completed") {
        stats.overdue++;
      }

      // Calculate on-time completion
      if (project.status === "completed" && project.completedDate) {
        completedCount++;
        const completionTime = project.completedDate - project.startDate;
        totalCompletionTime += completionTime;

        if (!project.dueDate || project.completedDate <= project.dueDate) {
          stats.completedOnTime++;
        }
      }

      // Sum budgets
      stats.totalBudget += project.budget || 0;
      stats.totalActualCost += project.actualCost || 0;
    }

    // Calculate average completion time in days
    if (completedCount > 0) {
      stats.averageCompletionTime = Math.round(totalCompletionTime / completedCount / (1000 * 60 * 60 * 24));
    }

    return stats;
  },
});

// Get project timeline
export const getProjectTimeline = query({
  args: { id: v.id("projects") },
  handler: async (ctx, args) => {
    const project = await ctx.db.get(args.id);
    if (!project) return null;

    // Get all activity logs for this project
    const activities = await ctx.db
      .query("activityLogs")
      .withIndex("by_entity", (q) =>
        q.eq("entityType", "project").eq("entityId", args.id)
      )
      .order("desc")
      .collect();

    // Get deliverables timeline
    const deliverables = await ctx.db
      .query("deliverables")
      .withIndex("by_project", (q) => q.eq("projectId", args.id))
      .collect();

    const timeline = [
      {
        date: project.createdAt,
        type: "project_created",
        title: "Project Created",
        description: `Project "${project.name}" was created`,
      },
      ...activities.map(activity => ({
        date: activity.timestamp,
        type: `project_${activity.action}`,
        title: `Project ${activity.action}`,
        description: JSON.stringify(activity.changes),
      })),
      ...deliverables.map(d => ({
        date: d.createdAt,
        type: "deliverable_added",
        title: `Deliverable Added: ${d.name}`,
        description: d.description,
      })),
    ];

    // Add milestones to timeline
    if (project.milestones) {
      project.milestones.forEach(m => {
        timeline.push({
          date: m.dueDate,
          type: "milestone",
          title: `Milestone: ${m.name}`,
          description: `Status: ${m.status}`,
        });
      });
    }

    // Sort timeline by date
    timeline.sort((a, b) => b.date - a.date);

    return timeline;
  },
});

// Create project from template
export const createProjectFromTemplate = mutation({
  args: {
    clientId: v.id("clients"),
    templateType: v.string(), // "website", "app", "automation"
    name: v.string(),
  },
  handler: async (ctx, args) => {
    const now = Date.now();

    // Define template configurations
    const templates: Record<string, any> = {
      website: {
        type: "website",
        milestones: [
          { name: "Discovery & Requirements", dueDate: now + 7 * 24 * 60 * 60 * 1000, status: "pending" },
          { name: "Design Mockups", dueDate: now + 14 * 24 * 60 * 60 * 1000, status: "pending" },
          { name: "Development", dueDate: now + 28 * 24 * 60 * 60 * 1000, status: "pending" },
          { name: "Testing & Launch", dueDate: now + 35 * 24 * 60 * 60 * 1000, status: "pending" },
        ],
        description: "Website development project",
        priority: "medium",
      },
      app: {
        type: "app",
        milestones: [
          { name: "Requirements Gathering", dueDate: now + 7 * 24 * 60 * 60 * 1000, status: "pending" },
          { name: "UI/UX Design", dueDate: now + 21 * 24 * 60 * 60 * 1000, status: "pending" },
          { name: "MVP Development", dueDate: now + 60 * 24 * 60 * 60 * 1000, status: "pending" },
          { name: "Beta Testing", dueDate: now + 75 * 24 * 60 * 60 * 1000, status: "pending" },
          { name: "Production Release", dueDate: now + 90 * 24 * 60 * 60 * 1000, status: "pending" },
        ],
        description: "Application development project",
        priority: "high",
      },
      automation: {
        type: "automation",
        milestones: [
          { name: "Process Analysis", dueDate: now + 3 * 24 * 60 * 60 * 1000, status: "pending" },
          { name: "Workflow Design", dueDate: now + 7 * 24 * 60 * 60 * 1000, status: "pending" },
          { name: "Implementation", dueDate: now + 14 * 24 * 60 * 60 * 1000, status: "pending" },
          { name: "Testing & Deployment", dueDate: now + 21 * 24 * 60 * 60 * 1000, status: "pending" },
        ],
        description: "Process automation project",
        priority: "medium",
      },
    };

    const template = templates[args.templateType];
    if (!template) throw new Error("Invalid template type");

    const projectId = await ctx.db.insert("projects", {
      clientId: args.clientId,
      name: args.name,
      ...template,
      status: "planning",
      startDate: now,
      dueDate: template.milestones[template.milestones.length - 1].dueDate,
      createdAt: now,
      updatedAt: now,
    });

    return projectId;
  },
});